pennylane/workflow/run.py

```diff
@@ -0,0 +1,317 @@
+    """
+    inner_execute = _make_inner_execute(device, inner_transform_program, config)
+
+    def inner_execute_with_empty_jac(tapes, **_):
+        return inner_execute(tapes), []
+
+    execute_fn = inner_execute_with_empty_jac
+
+    if config.use_device_gradient:
+
+        if config.grad_on_execution:
+
+            def wrap_execute_and_compute_derivatives(internal_tapes):
+                """A partial function wrapping the execute_and_compute_derivatives method of the device."""
+                numpy_tapes, _ = qml.transforms.convert_to_numpy_parameters(internal_tapes)
+                return device.execute_and_compute_derivatives(numpy_tapes, config)
+
+            execute_fn = wrap_execute_and_compute_derivatives
+            diff_method = None
+
+        else:
+
+            def execution_with_dummy_jac(internal_tapes):
+                """A wrapper around device.execute that returns an empty tuple for derivatives."""
+                numpy_tapes, _ = qml.transforms.convert_to_numpy_parameters(internal_tapes)
+                return device.execute(numpy_tapes, config), tuple()
+
+            execute_fn = execution_with_dummy_jac
+
+            def device_compute_derivatives(internal_tapes):
+                """A partial function wrapping the compute_derivatives method of the device."""
+                numpy_tapes, _ = qml.transforms.convert_to_numpy_parameters(internal_tapes)
+                return device.compute_derivatives(numpy_tapes, config)
+
+            diff_method = device_compute_derivatives
+
+    elif config.grad_on_execution is True:
+        raise ValueError("Gradient transforms cannot be used with grad_on_execution=True")
+
+    else:
+        diff_method = config.gradient_method
+
+    return execute_fn, diff_method
+
+
+def _construct_ml_execution_pipeline(config, device, inner_transform_program):
+    """Constructs the ML execution pipeline for all interfaces except TF_AUTOGRAPH.
+
+    This function determines the execution function (`execute_fn`) and the Jacobian product
+    class (`jpc`) required for gradient computations.
+
+    Args:
+        config (qml.devices.ExecutionConfig): resolved execution configuration
+        device (qml.devices.Device): a Pennylane device
+        inner_transform_program (qml.transforms.core.TransformProgram): the transformation applied to quantum tapes
+
+    Returns:
+        tuple: A tuple containing:
+            - `jpc`: jacobian product class for computing gradients
+            - `execute_fn`: the function to execute quantum tapes
+
+    Raises:
+        ValueError: If gradients are computed on execution (`grad_on_execution=True`).
+    """
+    inner_execute = _make_inner_execute(device, inner_transform_program, config)
+    cache = _cache_transform in inner_transform_program
+
+    execute_fn = inner_execute
+    jpc = None
+
+    if config.use_device_jacobian_product:
+        jpc = DeviceJacobianProducts(device, config)
```

Comment:
```suggestion
    if config.use_device_jacobian_product:
        return DeviceJacobianProducts(device, config), execute_fn
```
Might make sense to have early exit on each branch.

This is stylistic, but I generally kind of prefer it.  It means I don't have to think about that branch anymore.